Spigot-Server-Patches/0680-Remove-streams-from-SensorNearest.patch

```diff
@@ -0,0 +1,111 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Bjarne Koll <lynxplay101@gmail.com>
+Date: Wed, 3 Mar 2021 12:48:48 +0100
+Subject: [PATCH] Remove streams from SensorNearest
+
+The behavioural nearby sensors are validated every tick on the entities
+that registered the respective sensors and are therefore a good subject
+to performance improvements.
+
+More specifically this commit replaces the Stream#filter usage with
+ArrayList#removeIf as the removeIf method on an array list is heavily
+optimized towards a single internal array re-allocation without any
+further overhead on the removeIf call.
+
+The only negative of this change is the rather agressive diff these
+patches introduce as the methods are basically being reimplemented
+compared to the previous stream-based implementation.
+
+See: https://nipafx.dev/java-stream-performance/
+
+diff --git a/src/main/java/net/minecraft/server/SensorNearestItems.java b/src/main/java/net/minecraft/server/SensorNearestItems.java
+index 2e747158d48ab28ac1611990cc97aa4a9e30b30e..9353fb7c30a438d6d0697049e2ee7252b62a2aa3 100644
+--- a/src/main/java/net/minecraft/server/SensorNearestItems.java
++++ b/src/main/java/net/minecraft/server/SensorNearestItems.java
+@@ -21,18 +21,10 @@ public class SensorNearestItems extends Sensor<EntityInsentient> {
+         List<EntityItem> list = worldserver.a(EntityItem.class, entityinsentient.getBoundingBox().grow(8.0D, 4.0D, 8.0D), (entityitem) -> {
+             return true;
+         });
+-
+-        entityinsentient.getClass();
++        // Paper start - remove streams in favour of lists
+         list.sort(Comparator.comparingDouble(entityinsentient::h));
+-        Stream stream = list.stream().filter((entityitem) -> {
+-            return entityinsentient.i(entityitem.getItemStack());
+-        }).filter((entityitem) -> {
+-            return entityitem.a((Entity) entityinsentient, 9.0D);
+-        });
+-
+-        entityinsentient.getClass();
+-        Optional<EntityItem> optional = stream.filter(entityinsentient::hasLineOfSight).findFirst();
+-
+-        behaviorcontroller.setMemory(MemoryModuleType.NEAREST_VISIBLE_WANTED_ITEM, optional);
++        list.removeIf(entityItem -> !entityinsentient.i(entityItem.getItemStack()) || !entityItem.a(entityinsentient, 9.0D) || !entityinsentient.hasLineOfSight(entityItem));
++        behaviorcontroller.setMemory(MemoryModuleType.NEAREST_VISIBLE_WANTED_ITEM, list.isEmpty() ? Optional.empty() : Optional.of(list.get(0)));
```

Comment:
Why not just iterate over the list and pick the first matching entry?

Reply:
Right! Yeah similar to how the player sensor does it.

Reply:
Okay I hope the latest change implemented it. Please resolve the conversation if you are satisfied :)